samples/net/openthread/coap_client_ble/src/ble_utils.c

```diff
@@ -0,0 +1,180 @@
+#include <bluetooth/hci.h>
+#include <bluetooth/uuid.h>
+#include <logging/log.h>
+#include <settings/settings.h>
+
+#include "ble_utils.h"
+
+LOG_MODULE_REGISTER(ble_utils, CONFIG_BLE_UTILS_LOG_LEVEL);
+
+#define DEVICE_NAME CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN (sizeof(DEVICE_NAME) - 1)
+
+static struct bt_conn *current_conn;
+ble_connection_cb_t on_connect;
+ble_disconnection_cb_t on_disconnect;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, NUS_UUID_SERVICE),
+};
+
+static struct k_work on_connect_work;
+static struct k_work on_disconnect_work;
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+	if (err) {
+		LOG_ERR("Connection failed (err %u)", err);
+		return;
+	}
+
+	LOG_INF("Connected");
+	current_conn = bt_conn_ref(conn);
+
+	k_work_submit(&on_connect_work);
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	LOG_INF("Disconnected (reason %u)", reason);
+
+	if (current_conn) {
+		bt_conn_unref(current_conn);
+		current_conn = NULL;
+
+		k_work_submit(&on_disconnect_work);
+	}
+}
+
+static char *log_addr(struct bt_conn *conn)
+{
+	static char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	return log_strdup(addr);
+}
+
+static void __attribute__((unused))
+security_changed(struct bt_conn *conn, bt_security_t level,
+		 enum bt_security_err err)
+{
+	char *addr = log_addr(conn);
+
+	if (!err) {
+		LOG_INF("Security changed: %s level %u", addr, level);
+	} else {
+		LOG_INF("Security failed: %s level %u err %d", addr, level,
+			err);
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	COND_CODE_1(CONFIG_BT_SMP, (.security_changed = security_changed), ())
+};
+
+static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey)
+{
+	LOG_INF("Passkey for %s: %06u", log_addr(conn), passkey);
+}
+
+static void auth_cancel(struct bt_conn *conn)
+{
+	LOG_INF("Pairing cancelled: %s", log_addr(conn));
+}
+
+static void pairing_confirm(struct bt_conn *conn)
+{
+	bt_conn_auth_pairing_confirm(conn);
+
+	LOG_INF("Pairing confirmed: %s", log_addr(conn));
+}
+
+static void pairing_complete(struct bt_conn *conn, bool bonded)
+{
+	LOG_INF("Pairing completed: %s, bonded: %d", log_addr(conn), bonded);
+}
+
+static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason)
+{
+	LOG_INF("Pairing failed conn: %s, reason %d", log_addr(conn), reason);
+}
+
+static struct bt_conn_auth_cb conn_auth_callbacks = {
+	.passkey_display = auth_passkey_display,
+	.cancel = auth_cancel,
+	.pairing_confirm = pairing_confirm,
+	.pairing_complete = pairing_complete,
+	.pairing_failed = pairing_failed
+};
+
+static struct bt_gatt_nus_cb *set_nus_clbs(nus_received_cb_t on_nus_received,
```

Comment:
this function saves us one line of code in the calling function but adds extra 10 lines here. consider moving its content into the calling func.

Reply:
It has been done this way to decouple application dependencies. `ble_utils` code has been prepared to be separate module in case we need some other samples with BLE interaction. It contains common code for all BLE samples. It can be also very easly ported to tha client applications.